Better pretty-printer output.

liautaud opened the pull request:
This pull request solves two issues regarding the `Irmin.Type` pretty-printer:
1. Right now, the pretty printer falls back to the JSON encoder in some cases (see `src/irmin/type/type_pp.ml:27`). This can lead to consistency issues when the behavior of the JSON encoder changes, which would be the case if #979 gets merged for instance.
2. Right now, the output of the pretty-printer is a mixture of OCaml-like syntax and JSON. Ideally (cc. @CraigFe), the pretty-printer should "give us something that can be copy-pasted" into an OCaml REPL, which could be helpful when debugging Alcotest diffs.

The first commit rewrites `Type_pp.t` to remove the dependency on the JSON encoder, and to use a new output format which "can be copy-pasted". The second commit updates the unit tests to reflected the new (breaking) output of the pretty-printer.

CraigFe commented on src/irmin/type/type_pp.ml:
```diff
@@ -24,20 +24,61 @@ let t t =
+  and record : type a. a record -> a pp =
+   fun t ->
+    fields t
+    |> List.map (fun (Field f) -> Fmt.Dump.field f.fname f.fget (aux f.ftype))
```
```suggestion
    |> List.map (fun (Field f) -> Fmt.Dump.field ~label:Fmt.string f.fname f.fget (aux f.ftype))
```

Fmt inserts colours here by default for some reason.

CraigFe commented on test/irmin/test_type.ml:
```diff
@@ -251,34 +251,38 @@ let test_to_string () =
   test "record" my_record_t
     { foo = 2; flag = false; letter = Delta }
-    "{\"foo\":2,\"flag\":0,\"letter\":\"Delta\"}";
+    {|{ "foo" = 2;
+  "flag" = false;
+  "letter" = Delta }|};
```
Wrapping field names in `" ... "` looks off to me.

CraigFe commented on src/irmin/type/type_pp.ml:
```diff
@@ -24,20 +24,57 @@ let t t =
+    | Var v -> raise (Unbound_type_variable v)
+    | List l -> Fmt.Dump.list (aux l.v) ppf x
+    | Array a -> Fmt.Dump.array (aux a.v) ppf x
+    | Option o -> Fmt.Dump.option (aux o) ppf x
```
Will need to implement this by-hand to add parentheses (necessary for e.g. nested `Some`s). The test cases for `result` are doing this properly, and we should be symmetric here.

liautaud replied:
You're right. I've added similar test cases for `option`, and re-implemented option by-hand.
We could be a bit smarter and only parenthesize when needed, but it's probably too much effort for too little reward.

samoht commented on test/irmin/test_type.ml:
```diff
@@ -251,34 +251,38 @@ let test_to_string () =
   test "recursive record" my_recursive_record_t
     { head = 1; tail = Some { head = 2; tail = None } }
-    "{\"head\":1,\"tail\":{\"head\":2}}";
+    {|{ "head" = 1;
```
if you want the output of `pp` to be closer of the OCaml syntax you probably don't want double quotes around `"head"`.

CraigFe commented on test/irmin/test_type.ml:
```diff
@@ -226,21 +226,21 @@ let test_to_string () =
   test "float{NaN}" T.float Stdlib.nan "nan";
-  test "bytes" T.bytes (Bytes.make 5 'a') "aaaaa";
-  test "string" T.string "foo\nbar\\" "foo\nbar\\";
+  test "bytes" T.bytes (Bytes.make 5 'a') "\"aaaaa\"";
+  test "string" T.string "foo\nbar\\" "\"foo\\nbar\\\\\"";
```
I think escaping strings to be directly importable to OCaml is probably a step too far: I think it would just confuse me. Other thoughts? (CC @samoht)

samoht replied:
yea, no need to espace strings I think

samoht replied:
Actually after some thoughts I am not so sure. If the goal is to have an OCaml output (e.g. something similar to `Fmt.Dump`) better stay consistent.

CraigFe replied:
The problem from my perspective is that `Type.to_string` is directly derived from this pretty-printer, and I wouldn't expect such behaviour from a `to_string` function. Perhaps the answer is to have an optional `?escape_ocaml` that is set to false in the `to_string` case.

Having thought of that, this change probably introduces a large number of cases where `to_string >> of_string` is always an Error (since `of_string` is still delegating to `Type_json.of_string` in most complex cases.) I think that probably needs to be dealt with in one way or another.

liautaud replied:
I feel like not escaping would prove frustrating quite quickly–imagine having an "almost-ready-to-paste" representation, but still having to quote manually before you can actually paste. As for the "it would confuse me", I guess it really depends on context. A standalone string with quotes might be confusing, but a quoted string in the inside an option or a record feels natural enough to me.

liautaud replied:
Also, if we don't escape, we might get confused if we got something an output like `Some`, which could just as much be the constructor or the string `"Some"`.

CraigFe replied:
Fair points. The case for Bytes still applies since there is no bytes literal (and we're currently making them look like string literals...). Can we go back to the default behaviour there?

CraigFe replied:
> Also, if we don't escape, we might get confused if we got something an output like `Some`, which could just as much be the constructor or the string `"Some"`.

I think this confuses surrounding with quote marks (which I think is definitely good) with escaping the inner string (which I think is probably not, or is at least only useful in certain contexts). Again, I think it's fine for `pp` to do something like that, but definitely wouldn't expect it of `to_string`.

liautaud replied:
Oh, there might actually be a significant issue that I missed. After searching for `to_string` in the codebase quickly, it seems like the pretty-printer is used in a lot of places as a quick way to turn an "unknown" Irmin type into an unquoted string (e.g. in `irmin-git`, in `dot.ml` or in `commit.ml`).

What is `Irmin.Type.to_string` actually supposed to be used for, then?

liautaud replied:
> > Also, if we don't escape, we might get confused if we got something an output like `Some`, which could just as much be the constructor or the string `"Some"`.
> 
> I think this confuses surrounding with quote marks (which I think is definitely good) with escaping the inner string (which I think is probably not, or is at least only useful in certain contexts). Again, I think it's fine for `pp` to do something like that, but definitely wouldn't expect it of `to_string`.

Right now, `to_string` and `pp` are defined using one another (`let to_string ty = Fmt.to_to_string (pp ty)`).
Should we break this behavior, and separate the two?

CraigFe replied:
(As said above, I think the solution is to have `pp` take `?(escape_ocaml = true)` which is set to `false` when building the `to_string` function. Then `of_string` will need to implement the inverse of `pp ~escape_ocaml:true`.)

liautaud replied:
Would `escape_ocaml=false` still output quotes around strings, or would we get rid of them altogether to maintain backwards compatibility?

CraigFe replied:
I vote not, so that `Type.(to_string string)` is identity, but there are advantages both ways. I leave it up to you :slightly_smiling_face:

samoht commented:
My only worry with changing `pp` is that we need to make sure that this gives a useful result for CLI users. But maybe this doesn't matter as it's using `pp_json` directly?

liautaud commented:
After taking a quick look, it looks like the CLI is mostly using `pp`. Maybe we could switch it to `pp_json` if the OCaml syntax is really too confusing for CLI users, though?

liautaud commented:
After taking all the previous remarks into account, I've reverted most of the changes to `to_string` to maintain backwards compatibility, and I've added a `?ocaml_syntax:bool` optional argument to `pp` to enable the "copy-pasteable" output.

As a shorthand, I've added `to_ocaml_string` to the exports of `Type_pp`, so we now have:
```ocaml
let to_string ty = Fmt.to_to_string (t ~ocaml_syntax:false ty)
let to_ocaml_string ty = Fmt.to_to_string (t ~ocaml_syntax:true ty)
```

Finally, since `to_string` will continue to rely on the JSON encoder, I've turned the representation of "special" floats to strings to avoid issues when #979 gets merged. In other words, we now have:
```ocaml
test "float{NaN}" T.float Stdlib.nan "\"nan\"";
test "float{inf}" T.float Stdlib.infinity "\"inf\"";
test "float{-inf}" T.float Stdlib.neg_infinity "\"-inf\""; 
```

CraigFe commented on src/irmin/type/type.mli:
```diff
@@ -298,14 +298,17 @@ type 'a pp = 'a Fmt.t
 
-val pp : 'a t -> 'a pp
+val pp : ?ocaml_syntax:bool -> 'a t -> 'a pp
 (** [pp t] is the pretty-printer for values of type [t]. *)
```
Documentation for `~ocaml_syntax` needed here, including at least a specification of the default. (I'd be in favour of making it `?ocaml_syntax:unit` to remove that ambiguity altogether.)

CraigFe commented on src/irmin/type/type_pp.ml:
```diff
@@ -16,28 +16,78 @@
+    | Float, false when Float.is_nan x -> Fmt.string ppf "\"nan\""
+    | Float, false when x = infinity -> Fmt.string ppf "\"inf\""
+    | Float, false when x = neg_infinity -> Fmt.string ppf "\"-inf\""
+    | String _, false -> Fmt.string ppf x
+    | Unit, true -> Fmt.string ppf "()"
+    | Char, true -> Fmt.(pf ppf "'%c'" x)
```
Let's group `| Foo, false` and `| Foo, true` together in the ordering to keep things lexicographic.

liautaud commented:
@CraigFe Fixed the few nits.

CraigFe commented on test/irmin/test_type.ml:
```diff
@@ -325,6 +333,88 @@ let test_to_string () =
 
   ()
 
+(** Test the behaviour of {!Irmin.Type.to_ocaml_string}. *)
+let test_to_ocaml_string () =
```
```suggestion
(** Test the behaviour of {!Irmin.Type.pp_dump}. *)
let test_pp_dump () =
```

CraigFe commented on test/irmin/test_type.ml:
```diff
@@ -775,6 +865,7 @@ let suite =
     ("json_option", `Quick, test_json_option);
     ("bin", `Quick, test_bin);
     ("to_string", `Quick, test_to_string);
+    ("to_ocaml_string", `Quick, test_to_ocaml_string);
```
```suggestion
    ("pp_dump", `Quick, test_pp_dump);
```

CraigFe commented:
Thanks!
